src/hotspot/share/oops/symbolHandle.hpp

```diff
@@ -76,6 +110,23 @@ class SymbolHandleBase : public StackObj {
     Symbol::maybe_decrement_refcount(_temp);
   }
 
+  // Keep this symbol alive for some time to allow for reuse.
+  // Temp symbols for the same string can often be created in quick succession,
+  // and this queue allows them to be reused instead of churning.
+  void add_to_cleanup_delay_queue(Symbol* sym) {
+    TempSymbolDelayQueueNode* node = new TempSymbolDelayQueueNode(sym);
+    _cleanup_delay.push(*node);
+
+    // If the queue is now full, implement a one-in, one-out policy.
+    if (Atomic::add(&_cleanup_delay_len, 1, memory_order_relaxed) > _cleanup_delay_max_entries) {
+      TempSymbolDelayQueueNode* result = _cleanup_delay.pop();
+      if (result != nullptr) {
+        delete result;
+        Atomic::dec(&_cleanup_delay_len);
```

Comment:
Because of a limitation on NonblockingQueue (from the class description: "A
queue may appear empty even though elements have been added and not
removed."), it is theoretically possible for the max-entries value to be
exceeded.  (List is empty, thread1 starts a push but is paused, other threads
push lots of entries.)  But that will eventually be cleaned up by completion
of the initial push and then later draining the list.  So I don't think this
is a problem in practice, but wanted to note that I'd looked at the question.

Reply:
Thanks! Yes, I agree that it's not perfectly bounded but that this usage works out okay.